File: Sliding Window/76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        s_length, t_length = len(s), len(t) 

        if t_length > s_length:
            return "" 
        elif t == s:
            return s
        elif t_length == 1:
            try:
                s.index(t)
                return t
            except:
                return ""

        ans=""
        t_frequency = collections.Counter(t)
        track = dict()

        def test():
            for key, val in t_frequency.items():
                if val > track[key]:
                    return False
            return True

        #init
        for k in t_frequency.keys():
            track.update(
                {k : 0}
                )

        #first look
        left_ptr = 0
        right_ptr = 0
        for i in range(s_length): 
            if s[i] in t_frequency.keys():
                left_ptr = i
                break
        
        for i in range(left_ptr, s_length):
            
            #move right
            right_ptr = i
            if s[i] in t_frequency.keys():
                track.update({s[i]: track[s[i]] + 1})
                while test():
                    w=s[left_ptr: right_ptr + 1]
                    if ans=="" or len(w)<len(ans):
                        ans=w

                    #move left
                    track.update({s[left_ptr]: track[s[left_ptr]] - 1}) 
                    for j in range(left_ptr + 1, right_ptr + 1):
                        if s[j] in t_frequency.keys():
                            left_ptr = j
                            break

        if (test()):
            if ans == "" or len(w) < len(ans):
                ans = w

        return ans

File: Sliding Window/76-minimum-window-substring/minimum-window-substring.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        int s_length = s.length();
        int t_length = t.length();

        if (t_length > s_length) {
            return "";
        } else if (t == s) {
            return s;
        } 
        // else if (t_length == 1) {
        //     try {
        //         s.find(t);
        //         return t;
        //     } catch (...) {
        //         return "";
        //     }
        // }

        string ans = "";
        unordered_map<char, int> t_frequency;
        unordered_map<char, int> track;

        auto test = [&]() -> bool {
            for (const auto& entry : t_frequency) {
                if (entry.second > track[entry.first]) {
                    return false;
                }
            }
            return true;
        };

        // Initialize
        for (char c : t) {
            t_frequency[c]++;
            track[c] = 0;
        }

        int left_ptr = 0;
        int right_ptr = 0;
        int min_window_start = -1;
        int min_window_length = INT_MAX;

        for (int i = 0; i < s_length; ++i) {
            if (t_frequency.find(s[i]) != t_frequency.end()) {
                track[s[i]]++;
                while (test()) {
                    int current_window_length = i - left_ptr + 1;
                    if (current_window_length < min_window_length) {
                        min_window_length = current_window_length;
                        min_window_start = left_ptr;
                    }

                    // Move left
                    if (t_frequency.find(s[left_ptr]) != t_frequency.end()) {
                        track[s[left_ptr]]--;
                    }
                    left_ptr++;
                }
            }
        }

        if (min_window_start != -1) {
            ans = s.substr(min_window_start, min_window_length);
        }

        return ans;
    }
};
